### PR TITLE
Add fallback filter widgets for uncommon (or unknown by Metabase) data types

### LINKED ISSUE #45252: Reduce Filter Options to Those that Work (Array fields)
### Describe the bug

If you have an array column and you attempt to filter on it, most filter options are not currently supported: https://github.com/metabase/metabase/issues/18108

It would be ideal if the GUI didn't present filter options that don't currently work ([like we do with JSON and XML columns](https://github.com/metabase/metabase/pull/23872)).

### To Reproduce

1. Set up a test Metabase Instance
2. In Postgres set up a test table with arrays:

CREATE TABLE test_table (
    id SERIAL PRIMARY KEY,
    identifiers VARCHAR[] NOT NULL
);

INSERT INTO test_table (identifiers) values
    (ARRAY['abc', 'def', 'ghi']),
    (ARRAY['jkl', 'mno']),
    (ARRAY['pqr', 'stu', 'vwx', 'yz']),
    (ARRAY['test1', 'test2', 'test3']);

3. Interact with filters from table columns, the filter modal, etc
4. If the column has no semantic type set the filter options simply don't react when you click on them
5. If the column is set to a semantic type, like Category, the empty/not empty filters work but all other options in the menu will fail (see above linked issue)

### Expected behavior

We shouldn't present the user with unsupported filter options 

### Logs

_No response_

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.23+9",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.23",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.23+9",
    "os.name": "Linux",
    "os.version": "5.15.133.1-microsoft-standard-WSL2",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "snowflake",
      "postgres",
      "athena",
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.22 (Debian 11.22-1.pgdg110+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.2"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2024-06-15",
      "tag": "v1.49.17",
      "hash": "7e2b7bb"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

Confusing to users

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ranquild**
Fixes https://github.com/metabase/metabase/issues/45252

Previously we showed filter widgets only for known supported types (numbers, strings, etc). But we have other types in our type system that didn't have filter widgets at all. As we can always use `IS NULL` and `IS NOT NULL` filters, let's expose them in the UI. 

For the filter picker (notebook, ...) I used the same layout as for boolean filters. For the filter modal, I did something similar to boolean filters as well. The checkboxes were used because otherwise there would be no way of removing this filter in the filter modal.

How to verify:
- Follow the steps from the issue and create a PG table with an array field.
- Add this database to Metabase and try to add a filter for `Identifiers`. 

Also did some cleanup:
- Simplified `operators.cljc`
- Renamed `isDate` to `isTemporal` as in clojure because `isTemporal` is actually true for `Time`, so the name was incorrect
- Checked for allowed operators before using something as the default operator

Screenshot 2024-07-16 at 15 06 44
Screenshot 2024-07-16 at 15 06 46
Screenshot 2024-07-16 at 15 06 56

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/util.ts [hunk 1]
```diff
@@ -160,7 +160,7 @@ const getColumnExample = (
     return "2042-01-01";
   }
 
-  if (Lib.isDate(column)) {
+  if (Lib.isTemporal(column)) {
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/expressions/CombineColumns/util.ts:163]: This change clearly surfaces issues where we didn't support type/Time properly

File: frontend/src/metabase/querying/hooks/use-default-filter/use-default-filter.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  const [operator, setOperator] = useState(
+    filterParts
+      ? filterParts.operator
+      : getDefaultOperator(availableOptions, hasInitialOperator),
+  );
+
+  return {
+    operator,
+    availableOptions,
+    getFilterClause: (operator: Lib.DefaultFilterOperatorName | undefined) =>
```
Comment (kamilmielnik) [inline frontend/src/metabase/querying/hooks/use-default-filter/use-default-filter.ts:45]: `useCallback`?

File: frontend/src/metabase/querying/components/FilterPicker/DefaultFilterPicker/DefaultFilterPicker.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      />
+      <div>
+        <Radio.Group value={operator} onChange={handleOperatorChange}>
+          <Stack p="md" spacing="sm">
+            {availableOptions.map(option => (
+              <Radio
+                key={option.operator}
+                value={option.operator}
+                label={option.name}
+                pb={6}
```
Comment (kamilmielnik) [inline frontend/src/metabase/querying/components/FilterPicker/DefaultFilterPicker/DefaultFilterPicker.tsx:73]: Let's use `spacing="md"` on parent `<Stack>` instead

```suggestion
```

File: e2e/test/scenarios/filters-reproductions/filters-reproductions.cy.spec.js [hunk 4]
```diff
@@ -1200,6 +1206,62 @@ describe("issue 40622", () => {
...
+    getNotebookStep("filter")
+      .findByText("Add filters to narrow your answer")
+      .click();
+    popover().within(() => {
+      cy.findByText("Binary").scrollIntoView().click();
+      cy.findByLabelText("Is empty").click();
+      cy.button("Add filter").click();
+    });
+    visualize();
+    cy.wait("@dataset");
```
Comment (kamilmielnik) [inline e2e/test/scenarios/filters-reproductions/filters-reproductions.cy.spec.js:1236]: `visualize` already does `cy.wait("@dataset");`

```suggestion
    visualize();
```

**Review by kamilmielnik — APPROVED** [review]
:sweat_drops: :+1:

File: frontend/src/metabase/common/utils/columns.ts [hunk 5]
```diff
@@ -37,5 +37,5 @@ export function getColumnIcon(column: Lib.ColumnMetadata): IconName {
     return "int";
   }
 
-  return "unknown";
+  return "list";
```
Comment (kamilmielnik) [inline frontend/src/metabase/common/utils/columns.ts:40]: I think we can do better, let's ask design team for help.

Comment (ranquild) [reply frontend/src/metabase/common/utils/columns.ts:40]: That's what they suggested

**Pull request merged**